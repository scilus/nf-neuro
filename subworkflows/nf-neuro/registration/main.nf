include { REGISTRATION_ANATTODWI  } from '../../../modules/nf-neuro/registration/anattodwi/main'
include { REGISTRATION_ANTS   } from '../../../modules/nf-neuro/registration/ants/main'
include { REGISTRATION_EASYREG   } from '../../../modules/nf-neuro/registration/easyreg/main'
include { REGISTRATION_SYNTHREGISTRATION } from '../../../modules/nf-neuro/registration/synthregistration/main'

params.run_easyreg      = false
params.run_synthmorph   = false

workflow REGISTRATION {

    // The subworkflow requires at least ch_image and ch_ref as inputs to
    // properly perform the registration. Supplying a ch_metric will select
    // the REGISTRATION_ANATTODWI module meanwhile NOT supplying a ch_metric
    // will select the REGISTRATION_ANTS (SyN or SyNQuick) module. Alternatively,
    // NOT supplying ch_metric and activating alternative module flag with select
    // REGISTRATION_EASYREG or REGISTRATION_SYNTHMORPH

    take:
        ch_image                // channel: [ val(meta), image ]
        ch_ref                  // channel: [ val(meta), reference ]
        ch_metric               // channel: [ val(meta), metric ], optional
        ch_mask                 // channel: [ val(meta), mask ], optional
        ch_segmentation         // channel: [ val(meta), segmentation ], optional
        ch_ref_segmentation     // channel: [ val(meta), ref-segmentation ], optional

    main:

        ch_versions = Channel.empty()

        if ( params.run_easyreg ) {
            // ** Registration using Easyreg ** //
            // Result : [ meta, reference, image | [], ref-segmentation | [], segmentation | [] ]
            //  Steps :
            //   - join [ meta, reference, image | null ]
            //   - join [ meta, reference, image | null, ref-segmentation | null ]
            //   - join [ meta, reference, image | null, ref-segmentation | null, segmentation | null ]
            //   -  map [ meta, reference, image | [], ref-segmentation | [], segmentation | [] ]
            ch_register = ch_ref
                .join(ch_image, remainder: true)
                .join(ch_ref_segmentation, remainder: true)
                .join(ch_segmentation, remainder: true)
                .map{ it[0..1] + [it[2] ?: [], it[3] ?: [], it[4] ?: []] }

            REGISTRATION_EASYREG ( ch_register )
            ch_versions = ch_versions.mix(REGISTRATION_EASYREG.out.versions.first())

            // ** Set compulsory outputs ** //
            image_warped = REGISTRATION_EASYREG.out.flo_reg
            transfo_image = REGISTRATION_EASYREG.out.fwd_field
            transfo_trk = REGISTRATION_EASYREG.out.bak_field
            ref_warped = REGISTRATION_EASYREG.out.ref_reg

            // ** Set optional outputs. ** //
            // If segmentations are not provided as inputs,
            // easyreg will outputs synthseg segmentations
            out_segmentation = ch_segmentation.mix( REGISTRATION_EASYREG.out.flo_seg )
            out_ref_segmentation = ch_ref_segmentation.mix( REGISTRATION_EASYREG.out.ref_seg )
        }
        else if ( params.run_synthmorph ) {
            // ** Registration using synthmorph ** //
            ch_register = ch_image
                .join(ch_ref)

            REGISTRATION_SYNTHREGISTRATION ( ch_register )
            ch_versions = ch_versions.mix(REGISTRATION_SYNTHREGISTRATION.out.versions.first())

            // ** Set outputs ** //
            image_warped = REGISTRATION_SYNTHREGISTRATION.out.warped_image
            transfo_image = REGISTRATION_SYNTHREGISTRATION.out.warp
                .join(REGISTRATION_SYNTHREGISTRATION.out.affine)
            transfo_trk = Channel.empty()
            ref_warped = Channel.empty()
            out_segmentation = Channel.empty()
            out_ref_segmentation = Channel.empty()
        }
        else {
            // ** Classic registration using antsRegistration  ** //
            // Result : [ meta, image, reference, metric | [] ]
            //  Steps :
            //   - join [ meta, image, ref ]
            //   - join [ meta, image, ref, metric | null ]
            //   - map  [ meta, image, ref, metric | [] ]
            // Branches :
            //   - anat_to_dwi : has a metric at index 3
            //   - ants_syn    : doesn't have a metric at index 3 ( [] or null )
            ch_register = ch_image
                .join(ch_ref)
                .join(ch_metric, remainder: true)
                .map{ it[0..2] + [it[3] ?: []] }
                .branch{
                    anat_to_dwi : it[3]
                    ants_syn: true
                }

            // ** Registration using ANAT TO DWI ** //
            REGISTRATION_ANATTODWI ( ch_register.anat_to_dwi )
            ch_versions = ch_versions.mix(REGISTRATION_ANATTODWI.out.versions.first())

            // ** Set compulsory outputs ** //
            image_warped = REGISTRATION_ANATTODWI.out.t1_warped
            transfo_image = REGISTRATION_ANATTODWI.out.warp
                .join(REGISTRATION_ANATTODWI.out.affine)
            transfo_trk = REGISTRATION_ANATTODWI.out.affine
                .join(REGISTRATION_ANATTODWI.out.inverse_warp)

            // ** Registration using ANTS SYN SCRIPTS ** //
            // Registration using antsRegistrationSyN.sh or antsRegistrationSyNQuick.sh, has
            // to be defined in the config file or else the default (SyN) will be used.
            // Result : [ meta, image, mask | [] ]
            //  Steps :
            //   - join [ meta, image, metric | [], mask | null ]
            //   - map  [ meta, image, mask | [] ]
            ch_register = ch_register.ants_syn
                .join(ch_mask, remainder: true)
                .map{ it[0..2] + [it[4] ?: []] }

            REGISTRATION_ANTS ( ch_register )
            ch_versions = ch_versions.mix(REGISTRATION_ANTS.out.versions.first())

            // ** Set compulsory outputs ** //
            image_warped = image_warped.mix(REGISTRATION_ANTS.out.image)
            transfo_image = transfo_image.mix(REGISTRATION_ANTS.out.warp)
                .join(REGISTRATION_ANTS.out.affine)
            transfo_trk = transfo_trk.mix(REGISTRATION_ANTS.out.inverse_affine)
                .join(REGISTRATION_ANTS.out.inverse_warp)

            // **et optional outputs **//
            ref_warped = Channel.empty()
            out_segmentation = Channel.empty()
            out_ref_segmentation = Channel.empty()
        }

    emit:
        image_warped        = image_warped            // channel: [ val(meta), image ] ]
        ref_warped          = ref_warped              // channel: [ val(meta), ref ]
        transfo_image       = transfo_image           // channel: [ val(meta), [ warp ], [ <affine> ] ]
        transfo_trk         = transfo_trk             // channel: [ val(meta), [ <inverse-affine> ], [ inverse-warp ] ]
        segmentation        = out_segmentation        // channel: [ val(meta), segmentation ]
        ref_segmentation    = out_ref_segmentation    // channel: [ val(meta), ref-segmentation ]
        versions            = ch_versions             // channel: [ versions.yml ]
}
