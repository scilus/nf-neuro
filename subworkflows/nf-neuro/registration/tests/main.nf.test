nextflow_workflow {

    name "Test Subworkflow REGISTRATION"
    script "../main.nf"
    workflow "REGISTRATION"

    tag "subworkflows"
    tag "subworkflows_nfcore"
    tag "subworkflows/registration"

    tag "registration"
    tag "registration/anattodwi"
    tag "registration"
    tag "registration/ants"
    tag "registration/easyreg"

    tag "load_test_data"

    setup {
        run("LOAD_TEST_DATA", alias: "LOAD_DATA") {
            script "../../load_test_data/main.nf"
            process {
                """
                input[0] = Channel.from( [ "T1w.zip", "b0.zip", "dti.zip" ] )
                input[1] = "test.load-test-data"
                """
            }
        }
    }

    test("registration - antsRegistration") {
        config "./nextflow.config"
        when {
            workflow {
                """
                ch_split_test_data = LOAD_DATA.out.test_data_directory
                    .branch{
                        t1w: it.simpleName == "T1w"
                        b0: it.simpleName == "b0"
                        dti: it.simpleName == "dti"
                    }
                input[0] = ch_split_test_data.t1w.map{
                    test_data_directory -> [
                        [ id:'test', single_end:false ],
                        file("\${test_data_directory}/T1w.nii.gz")
                    ]
                }
                input[1] = ch_split_test_data.b0.map{
                    test_data_directory -> [
                        [ id:'test', single_end:false ],
                        file("\${test_data_directory}/b0.nii.gz")
                    ]
                }
                input[2] = ch_split_test_data.dti.map{
                    test_data_directory -> [
                        [ id:'test', single_end:false ],
                        file("\${test_data_directory}/fa.nii.gz")
                    ]
                }
                input[3] = Channel.empty()
                input[4] = Channel.empty()
                input[5] = Channel.empty()
                """
            }
        }

        then {
            assertAll(
                { assert workflow.success},
                { assert snapshot(workflow.out).match()}
            )
        }
    }

    test("registration - SyNQuick") {
        config "./nextflow.config"
        when {
            workflow {
                """
                ch_split_test_data = LOAD_DATA.out.test_data_directory
                    .branch{
                        t1w: it.simpleName == "T1w"
                        b0: it.simpleName == "b0"
                        dti: it.simpleName == "dti"
                    }
                input[0] = ch_split_test_data.t1w.map{
                    test_data_directory -> [
                        [ id:'test', single_end:false ],
                        file("\${test_data_directory}/T1w.nii.gz")
                    ]}
                input[1] = ch_split_test_data.b0.map{
                    test_data_directory -> [
                        [ id:'test', single_end:false ],
                        file("\${test_data_directory}/b0.nii.gz")
                    ]}
                input[2] = []
                input[3] = Channel.empty()
                input[4] = Channel.empty()
                input[5] = Channel.empty()
                """
            }
        }

        then {
            assertAll(
                { assert workflow.success},
                { assert snapshot(workflow.out).match()}
            )
        }
    }

    test("registration - easyreg") {
        config "./nextflow_easyreg.config"
        when {
            workflow {
                """
                ch_split_test_data = LOAD_DATA.out.test_data_directory
                    .branch{
                        t1w: it.simpleName == "T1w"
                        b0: it.simpleName == "b0"
                        dti: it.simpleName == "dti"
                    }
                input[0] = ch_split_test_data.t1w.map{
                    test_data_directory -> [
                        [ id:'test' ],
                        file("\${test_data_directory}/T1w.nii.gz")
                    ]
                }
                input[1] = ch_split_test_data.b0.map{
                    test_data_directory -> [
                        [ id:'test' ],
                        file("\${test_data_directory}/b0.nii.gz")
                    ]
                }
                input[2] = Channel.empty()
                input[3] = Channel.empty()
                input[4] = Channel.empty()
                input[5] = Channel.empty()
                """
            }
        }

        then {
            assertAll(
                { assert workflow.success},
                { assert snapshot(
                    file(workflow.out.image_warped.get(0).get(1)).name,
                    file(workflow.out.ref_warped.get(0).get(1)).name,
                    file(workflow.out.transfo_image.get(0).get(1)).name,
                    file(workflow.out.transfo_trk.get(0).get(1)).name,
                    niftiMD5SUM(workflow.out.segmentation.get(0).get(1), 6),
                    file(workflow.out.ref_segmentation.get(0).get(1)).name,
                    workflow.out.versions
                ).match()}
            )
        }
    }
}
