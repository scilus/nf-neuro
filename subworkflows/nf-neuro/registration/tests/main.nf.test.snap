{
    "registration - easyreg": {
        "content": [
            "test_floating_registered.nii.gz",
            "test_reference_registered.nii.gz",
            "test_forward_field.nii.gz",
            "test_backward_field.nii.gz",
            "test_floating_segmentation.nii.gz:md5:header,74ed7e917b9fde7b2aaba7f512cc1edf,data,e1ed38ea961180d3e4c03a56c1c7bc4c",
            "test_reference_segmentation.nii.gz",
            [
                "versions.yml:md5,3f38c911476c605480c02e484fcf6e9f"
            ]
        ],
        "meta": {
            "nf-test": "0.9.0",
            "nextflow": "24.04.4"
        },
        "timestamp": "2024-10-25T14:12:29.150266819"
    },
    "registration - antsRegistration": {
        "content": [
            {
                "0": [
                    [
                        {
                            "id": "test",
                            "single_end": false
                        },
                        "test__t1_warped.nii.gz:md5,9c34f424900adea263e39ed1f37a793f"
                    ]
                ],
                "1": [
                    
                ],
                "2": [
                    [
                        {
                            "id": "test",
                            "single_end": false
                        },
                        "test__output1Warp.nii.gz:md5,e0995852e251bee5521572a4495a0bd6",
                        "test__output0GenericAffine.mat:md5,6eef07f9004fb96e57ae0404ffef174c"
                    ]
                ],
                "3": [
                    [
                        {
                            "id": "test",
                            "single_end": false
                        },
                        "test__output0GenericAffine.mat:md5,6eef07f9004fb96e57ae0404ffef174c",
                        "test__output1InverseWarp.nii.gz:md5,28694d2b41f5209ab8f1acbb219014fe"
                    ]
                ],
                "4": [
                    
                ],
                "5": [
                    
                ],
                "6": [
                    "versions.yml:md5,df4f87e4cf013e11f65c490298bc1249"
                ],
                "image_warped": [
                    [
                        {
                            "id": "test",
                            "single_end": false
                        },
                        "test__t1_warped.nii.gz:md5,9c34f424900adea263e39ed1f37a793f"
                    ]
                ],
                "ref_segmentation": [
                    
                ],
                "ref_warped": [
                    
                ],
                "segmentation": [
                    
                ],
                "transfo_image": [
                    [
                        {
                            "id": "test",
                            "single_end": false
                        },
                        "test__output1Warp.nii.gz:md5,e0995852e251bee5521572a4495a0bd6",
                        "test__output0GenericAffine.mat:md5,6eef07f9004fb96e57ae0404ffef174c"
                    ]
                ],
                "transfo_trk": [
                    [
                        {
                            "id": "test",
                            "single_end": false
                        },
                        "test__output0GenericAffine.mat:md5,6eef07f9004fb96e57ae0404ffef174c",
                        "test__output1InverseWarp.nii.gz:md5,28694d2b41f5209ab8f1acbb219014fe"
                    ]
                ],
                "versions": [
                    "versions.yml:md5,df4f87e4cf013e11f65c490298bc1249"
                ]
            }
        ],
        "meta": {
            "nf-test": "0.9.0",
            "nextflow": "24.04.4"
        },
        "timestamp": "2024-10-25T09:25:14.344251"
    },
    "registration - SyNQuick": {
        "content": [
            {
                "0": [
                    [
                        {
                            "id": "test",
                            "single_end": false
                        },
                        "test__warped.nii.gz:md5,40010055cb124e9a5e61a2565f1f5db2"
                    ]
                ],
                "1": [
                    
                ],
                "2": [
                    [
                        {
                            "id": "test",
                            "single_end": false
                        },
                        [
                            "test__output0Warp.nii.gz:md5,790b8a076b1c146ffbb1d3ae614927d9",
                            "test__output1GenericAffine.mat:md5,b8f7939863fdef61b97d93f2b6021e12"
                        ]
                    ]
                ],
                "3": [
                    [
                        {
                            "id": "test",
                            "single_end": false
                        },
                        [
                            "test__output0InverseAffine.mat:md5,966aac28b5eda0fc6299672bfbeeb3ee",
                            "test__output1InverseWarp.nii.gz:md5,f6c7df8d65a8e31af3250bf8be5aece9"
                        ]
                    ]
                ],
                "4": [
                    
                ],
                "5": [
                    
                ],
                "6": [
                    "versions.yml:md5,9fe7f597ac6c76efff8851f9b672f17b"
                ],
                "image_warped": [
                    [
                        {
                            "id": "test",
                            "single_end": false
                        },
                        "test__warped.nii.gz:md5,40010055cb124e9a5e61a2565f1f5db2"
                    ]
                ],
                "ref_segmentation": [
                    
                ],
                "ref_warped": [
                    
                ],
                "segmentation": [
                    
                ],
                "transfo_image": [
                    [
                        {
                            "id": "test",
                            "single_end": false
                        },
                        [
                            "test__output0Warp.nii.gz:md5,790b8a076b1c146ffbb1d3ae614927d9",
                            "test__output1GenericAffine.mat:md5,b8f7939863fdef61b97d93f2b6021e12"
                        ]
                    ]
                ],
                "transfo_trk": [
                    [
                        {
                            "id": "test",
                            "single_end": false
                        },
                        [
                            "test__output0InverseAffine.mat:md5,966aac28b5eda0fc6299672bfbeeb3ee",
                            "test__output1InverseWarp.nii.gz:md5,f6c7df8d65a8e31af3250bf8be5aece9"
                        ]
                    ]
                ],
                "versions": [
                    "versions.yml:md5,9fe7f597ac6c76efff8851f9b672f17b"
                ]
            }
        ],
        "meta": {
            "nf-test": "0.9.0",
            "nextflow": "24.04.4"
        },
        "timestamp": "2024-10-25T09:19:43.397585"
    }
}